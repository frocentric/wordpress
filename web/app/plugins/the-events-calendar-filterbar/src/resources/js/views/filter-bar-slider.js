/* global tribe, jQuery, Swiper */
/* eslint-disable no-new, no-var, strict */

/**
 * Makes sure we have all the required levels on the Tribe Object.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar = tribe.filterBar || {};

/**
 * Configures Filter Bar Slider Object in the Global Tribe variable.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar.filterBarSlider = {};

/**
 * Initializes in a Strict env the code that manages the filter bar slider.
 *
 * @since 5.0.0
 *
 * @param  {PlainObject} $   jQuery
 * @param  {PlainObject} obj tribe.filterBar.filterBarSlider
 *
 * @return {void}
 */
( function( $, obj ) {
	'use strict';
	var $document = $( document );

	/**
	 * Selectors used for configuration and setup.
	 *
	 * @since 5.0.0
	 *
	 * @type {PlainObject}
	 */
	obj.selectors = {
		filterBarHorizontal: '.tribe-events--filter-bar-horizontal',
		filterBar: '.tribe-filter-bar',
		filtersSliderContainer: '[data-js="tribe-filter-bar-filters-slider-container"]',
		filtersSliderNav: '.tribe-filter-bar__filters-slider-nav',
		filtersSliderNavOverflowStart: '.tribe-filter-bar__filters-slider-nav--overflow-start',
		filtersSliderNavOverflowEnd: '.tribe-filter-bar__filters-slider-nav--overflow-end',
		filtersSliderNavButtonPrev: '[data-js~="tribe-filter-bar-filters-slider-nav-button-prev"]',
		filtersSliderNavButtonNext: '[data-js~="tribe-filter-bar-filters-slider-nav-button-next"]',
		filtersContainer: '[data-js="tribe-filter-bar-filters-container"]',
	};

	/**
	 * Handler for slider translate event.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.handleSliderOverflow = function( $container ) {
		var $filterBar = $container.find( obj.selectors.filterBar );
		var $filtersSliderContainer = $filterBar.find( obj.selectors.filtersSliderContainer );
		var $filtersSliderNav = $filtersSliderContainer.find( obj.selectors.filtersSliderNav );

		$filtersSliderNav
			.removeClass( obj.selectors.filtersSliderNavOverflowStart.className() )
			.removeClass( obj.selectors.filtersSliderNavOverflowEnd.className() );

		var swiper = $filtersSliderContainer[ 0 ].swiper;
		if ( ! swiper.isBeginning ) {
			$filtersSliderNav.addClass( obj.selectors.filtersSliderNavOverflowStart.className() );
		}
		if ( ! swiper.isEnd ) {
			$filtersSliderNav.addClass( obj.selectors.filtersSliderNavOverflowEnd.className() );
		}
	};

	/**
	 * Handler for slider translate event.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 * @param  {Number} translate  Value of slider translation.
	 *
	 * @return {void}
	 */
	obj.handleSliderTranslate = function( $container, translate ) {
		var $filterBar = $container.find( obj.selectors.filterBar );
		tribe.filterBar.filterToggle.closeAllFilters( $filterBar );

		var $filtersContainer = $filterBar.find( obj.selectors.filtersContainer );
		$filtersContainer.css( { transform: 'translateX(' + translate + 'px)' } );
	};

	/**
	 * Deinitializes filter bar slider.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.deinitSlider = function( $container ) {
		var $filtersSliderContainer = $container.find( obj.selectors.filtersSliderContainer );

		// If slider container doesn't exist, return early.
		if ( ! $filtersSliderContainer.length ) {
			return;
		}

		// If slider does not exist, return early.
		if ( ! $filtersSliderContainer[ 0 ].swiper ) {
			return;
		}

		// Deinitialize slider.
		var $filtersContainer = $container.find( obj.selectors.filtersContainer );
		$filtersSliderContainer[ 0 ].swiper.destroy();
		$filtersContainer.css( { transform: '' } );

		// Remove classes for overflow.
		$filtersSliderContainer
			.find( obj.selectors.filtersSliderNav )
			.removeClass( obj.selectors.filtersSliderNavOverflowStart.className() )
			.removeClass( obj.selectors.filtersSliderNavOverflowEnd.className() );
	};

	/**
	 * Initializes filter bar slider.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.initSlider = function( $container ) {
		var $filtersSliderContainer = $container.find( obj.selectors.filtersSliderContainer );

		// If slider container doesn't exist, return early.
		if ( ! $filtersSliderContainer.length ) {
			return;
		}

		// If slider exists, update and return early.
		if ( $filtersSliderContainer[ 0 ].swiper ) {
			$filtersSliderContainer[ 0 ].swiper.update();
			return;
		}

		// Initalize slider.
		new Swiper( $filtersSliderContainer[ 0 ], {
			slidesPerView: 'auto',
			resistanceRatio: 0,
			freeMode: true,
			freeModeMomentumBounce: false,
			navigation: {
				prevEl: $filtersSliderContainer.find( obj.selectors.filtersSliderNavButtonPrev )[ 0 ],
				nextEl: $filtersSliderContainer.find( obj.selectors.filtersSliderNavButtonNext )[ 0 ],
			},
			on: {
				init: function() {
					obj.handleSliderOverflow( $container );
				},
				setTranslate: function( translate ) {
					obj.handleSliderTranslate( $container, translate );
				},
				fromEdge: function() {
					obj.handleSliderOverflow( $container );
				},
				toEdge: function() {
					obj.handleSliderOverflow( $container );
				},
			},
		} );
	};

	/**
	 * Handler for resize event.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object of click event.
	 *
	 * @return {void}
	 */
	obj.handleResize = function( event ) {
		var $container = event.data.container;
		var containerState = $container.data( 'tribeEventsState' );
		var isMobile = containerState.isMobile;

		if ( isMobile ) {
			obj.deinitSlider( $container );
		} else {
			obj.initSlider( $container );
		}
	};

	/**
	 * Unbind events for filter bar slider functionality.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.unbindEvents = function( $container ) {
		$container.off( 'resize.tribeEvents', obj.handleResize );
	};

	/**
	 * Bind events for filter bar slider functionality.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.bindEvents = function( $container ) {
		$container.on( 'resize.tribeEvents', { container: $container }, obj.handleResize );
	};

	/**
	 * Deinitialize filter bar slider JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event} event event object for 'beforeAjaxSuccess.tribeEvents' event.
	 *
	 * @return {void}
	 */
	obj.deinit = function( event ) {
		var $container = event.data.container;
		obj.deinitSlider( $container );
		obj.unbindEvents( $container );
		$container.off( 'beforeAjaxSuccess.tribeEvents', obj.deinit );
	};

	/**
	 * Initialize filter bar slider JS.
	 *
	 * @since  5.0.0
	 *
	 * @param  {Event}   event      event object for 'afterSetup.tribeEvents' event.
	 * @param  {integer} index      jQuery.each index param from 'afterSetup.tribeEvents' event.
	 * @param  {jQuery}  $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.init = function( event, index, $container ) {
		// Return early if filter bar is not horizontal.
		if ( ! $container.is( obj.selectors.filterBarHorizontal ) ) {
			return;
		}

		obj.handleResize( { data: { container: $container } } );
		obj.bindEvents( $container );
		$container.on( 'beforeAjaxSuccess.tribeEvents', { container: $container }, obj.deinit );
	};

	/**
	 * Handles the initialization of filter bar slider when Document is ready.
	 *
	 * @since 5.0.0
	 *
	 * @return {void}
	 */
	obj.ready = function() {
		$document.on( 'afterSetup.tribeEvents', tribe.events.views.manager.selectors.container, obj.init );
	};

	// Configure on document ready.
	$( obj.ready );
} )( jQuery, tribe.filterBar.filterBarSlider );
