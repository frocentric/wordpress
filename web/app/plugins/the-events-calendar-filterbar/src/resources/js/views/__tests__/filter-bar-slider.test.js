describe( 'Filter Bar Slider', () => {
	beforeAll( () => {
		String.prototype.className = function() {
			if (
				(
					'string' !== typeof this &&
					! this instanceof String /* eslint-disable-line no-unsafe-negation */
				) ||
				'function' !== typeof this.replace
			) {
				return this;
			}

			return this.replace( '.', '' );
		};

		global.tribe = {};
		require( '../filter-bar-slider' );
		tribe.filterBar.filterToggle = {};
	} );

	afterAll( () => {
		delete String.prototype.className;
		delete global.tribe;
	} );

	describe( 'Selectors', () => {
		test( 'Should match snapshot', () => {
			const selectors = JSON.stringify( tribe.filterBar.filterBarSlider.selectors );
			expect( selectors ).toMatchSnapshot();
		} );
	} );

	describe( 'Handle slider overflow', () => {
		let $container;
		let $filterBar;
		let filtersSliderContainer;
		let $filtersSliderNav;

		beforeEach( () => {
			$container = $();
			$container.find = () => $filterBar;
			$filterBar = $();
			filtersSliderContainer = document.createElement( 'div' );
			filtersSliderContainer.swiper = {};
			$filtersSliderNav = $();
			$filtersSliderNav.addClass = jest.fn( () => $filtersSliderNav );
			$filtersSliderNav.removeClass = jest.fn( () => $filtersSliderNav );
		} );

		test( 'Should remove all overflow classes', () => {
			// Setup test.
			filtersSliderContainer.swiper.isBeginning = true;
			filtersSliderContainer.swiper.isEnd = true;
			const $filtersSliderContainer = $( filtersSliderContainer );
			$filtersSliderContainer.find = () => $filtersSliderNav;
			$filterBar.find = () => $filtersSliderContainer;

			// Test.
			tribe.filterBar.filterBarSlider.handleSliderOverflow( $container );

			// Confirm final states.
			expect( $filtersSliderNav.removeClass.mock.calls.length ).toBe( 2 );
			expect( $filtersSliderNav.addClass.mock.calls.length ).toBe( 0 );
		} );

		test( 'Should add overflow start class', () => {
			// Setup test.
			filtersSliderContainer.swiper.isBeginning = false;
			filtersSliderContainer.swiper.isEnd = true;
			const $filtersSliderContainer = $( filtersSliderContainer );
			$filtersSliderContainer.find = () => $filtersSliderNav;
			$filterBar.find = () => $filtersSliderContainer;

			// Test.
			tribe.filterBar.filterBarSlider.handleSliderOverflow( $container );

			// Confirm final states.
			expect( $filtersSliderNav.removeClass.mock.calls.length ).toBe( 2 );
			expect( $filtersSliderNav.addClass.mock.calls.length ).toBe( 1 );
			expect( $filtersSliderNav.addClass.mock.calls[ 0 ][ 0 ] )
				.toBe( tribe.filterBar.filterBarSlider.selectors.filtersSliderNavOverflowStart.className() );
		} );

		test( 'Should add overflow end class', () => {
			// Setup test.
			filtersSliderContainer.swiper.isBeginning = true;
			filtersSliderContainer.swiper.isEnd = false;
			const $filtersSliderContainer = $( filtersSliderContainer );
			$filtersSliderContainer.find = () => $filtersSliderNav;
			$filterBar.find = () => $filtersSliderContainer;

			// Test.
			tribe.filterBar.filterBarSlider.handleSliderOverflow( $container );

			// Confirm final states.
			expect( $filtersSliderNav.removeClass.mock.calls.length ).toBe( 2 );
			expect( $filtersSliderNav.addClass.mock.calls.length ).toBe( 1 );
			expect( $filtersSliderNav.addClass.mock.calls[ 0 ][ 0 ] )
				.toBe( tribe.filterBar.filterBarSlider.selectors.filtersSliderNavOverflowEnd.className() );
		} );

		test( 'Should add both overflow start and end classes', () => {
			// Setup test.
			filtersSliderContainer.swiper.isBeginning = false;
			filtersSliderContainer.swiper.isEnd = false;
			const $filtersSliderContainer = $( filtersSliderContainer );
			$filtersSliderContainer.find = () => $filtersSliderNav;
			$filterBar.find = () => $filtersSliderContainer;

			// Test.
			tribe.filterBar.filterBarSlider.handleSliderOverflow( $container );

			// Confirm final states.
			expect( $filtersSliderNav.removeClass.mock.calls.length ).toBe( 2 );
			expect( $filtersSliderNav.addClass.mock.calls.length ).toBe( 2 );
			expect( $filtersSliderNav.addClass.mock.calls[ 0 ][ 0 ] )
				.toBe( tribe.filterBar.filterBarSlider.selectors.filtersSliderNavOverflowStart.className() );
			expect( $filtersSliderNav.addClass.mock.calls[ 1 ][ 0 ] )
				.toBe( tribe.filterBar.filterBarSlider.selectors.filtersSliderNavOverflowEnd.className() );
		} );
	} );

	describe( 'Handle slider translate', () => {
		test( 'Should update filters container styles', () => {
			// Setup test.
			tribe.filterBar.filterToggle.closeAllFilters = jest.fn();
			const $filtersContainer = $( '<div></div>' );
			const $filterBar = $();
			$filterBar.find = () => $filtersContainer;
			const $container = $();
			$container.find = () => $filterBar;

			// Confirm initial state.
			expect( $filtersContainer.css( 'transform' ) ).toMatchSnapshot();

			// Test.
			tribe.filterBar.filterBarSlider.handleSliderTranslate( $container, 80 );

			// Confirm final states.
			expect( $filtersContainer.css( 'transform' ) ).toMatchSnapshot();
			expect( tribe.filterBar.filterToggle.closeAllFilters.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Deinitialize slider', () => {
		test( 'Should deinitialize filter slider', () => {
			// Setup test.
			const filtersSliderContainer = document.createElement( 'div' );
			filtersSliderContainer.swiper = {
				destroy: jest.fn(),
			};
			const $filtersSliderContainer = $( filtersSliderContainer );
			const $filtersContainer = $( '<div></div>' );
			const $container = $();
			$container.find = ( selector ) => {
				switch ( selector ) {
					case tribe.filterBar.filterBarSlider.selectors.filtersSliderContainer:
						return $filtersSliderContainer;
					case tribe.filterBar.filterBarSlider.selectors.filtersContainer:
						return $filtersContainer;
					default:
						return $();
				}
			};

			// Test.
			tribe.filterBar.filterBarSlider.deinitSlider( $container );

			// Confirm final state.
			expect( filtersSliderContainer.swiper.destroy.mock.calls.length ).toBe( 1 );
			expect( $filtersContainer.css( 'transform' ) ).toMatchSnapshot();
		} );
	} );

	describe( 'Initialize slider', () => {
		beforeEach( () => {
			global.Swiper = jest.fn();
		} );

		afterEach( () => {
			delete global.Swiper;
		} );

		test( 'Should initialize filter slider', () => {
			// Setup test.
			const $filtersSliderContainer = $( '<div></div>' );
			const $container = $();
			$container.find = () => $filtersSliderContainer;

			// Test.
			tribe.filterBar.filterBarSlider.initSlider( $container );

			// Confirm final state.
			expect( global.Swiper.mock.calls.length ).toBe( 1 );
		} );

		test( 'Should update and return early', () => {
			// Setup test.
			const filtersSliderContainer = document.createElement( 'div' );
			filtersSliderContainer.swiper = {
				update: jest.fn(),
			};
			const $filtersSliderContainer = $( filtersSliderContainer );
			const $container = $();
			$container.find = () => $filtersSliderContainer;

			// Test.
			tribe.filterBar.filterBarSlider.initSlider( $container );

			// Confirm final state.
			expect( global.Swiper.mock.calls.length ).toBe( 0 );
			expect( filtersSliderContainer.swiper.update.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Handle resize', () => {
		let deinitSliderHold;
		let initSliderHold;

		beforeEach( () => {
			deinitSliderHold = tribe.filterBar.filterBarSlider.deinitSlider;
			initSliderHold = tribe.filterBar.filterBarSlider.initSlider;
			tribe.filterBar.filterBarSlider.deinitSlider = jest.fn();
			tribe.filterBar.filterBarSlider.initSlider = jest.fn();
		} );

		afterEach( () => {
			tribe.filterBar.filterBarSlider.deinitSlider = deinitSliderHold;
			tribe.filterBar.filterBarSlider.initSlider = initSliderHold;
		} );

		test( 'Should deinitialize slider on mobile', () => {
			// Setup test.
			const $container = $( '<div></div>' );
			$container.data( 'tribeEventsState', { isMobile: true } );
			const event = {
				data: {
					container: $container,
				},
			};

			// Test.
			tribe.filterBar.filterBarSlider.handleResize( event );

			// Confirm final state.
			expect( tribe.filterBar.filterBarSlider.deinitSlider.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filterBarSlider.initSlider.mock.calls.length ).toBe( 0 );
		} );

		test( 'Should initialize slider on desktop', () => {
			// Setup test.
			const $container = $( '<div></div>' );
			$container.data( 'tribeEventsState', { isMobile: false } );
			const event = {
				data: {
					container: $container,
				},
			};

			// Test.
			tribe.filterBar.filterBarSlider.handleResize( event );

			// Confirm final state.
			expect( tribe.filterBar.filterBarSlider.deinitSlider.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filterBarSlider.initSlider.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Unbind events', () => {
		test( 'Should unbind events', () => {
			// Setup test.
			const $container = $();
			$container.off = jest.fn();

			// Test.
			tribe.filterBar.filterBarSlider.unbindEvents( $container );

			// Confirm final state.
			expect( $container.off.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Bind events', () => {
		test( 'Should bind events', () => {
			// Setup test.
			const $container = $();
			$container.on = jest.fn();

			// Test.
			tribe.filterBar.filterBarSlider.bindEvents( $container );

			// Confirm final state.
			expect( $container.on.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Deinitialize', () => {
		test( 'Should deinitialize filter bar slider', () => {
			// Setup test.
			const deinitSliderHold = tribe.filterBar.filterBarSlider.deinitSlider;
			const unbindEventsHold = tribe.filterBar.filterBarSlider.unbindEvents;
			tribe.filterBar.filterBarSlider.deinitSlider = jest.fn();
			tribe.filterBar.filterBarSlider.unbindEvents = jest.fn();
			const event = {
				data: {
					container: $(),
				},
			};

			// Test.
			tribe.filterBar.filterBarSlider.deinit( event, {}, {} );

			// Confirm final state.
			expect( tribe.filterBar.filterBarSlider.deinitSlider.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filterBarSlider.unbindEvents.mock.calls.length ).toBe( 1 );

			// Cleanup test.
			tribe.filterBar.filterBarSlider.deinitSlider = deinitSliderHold;
			tribe.filterBar.filterBarSlider.unbindEvents = unbindEventsHold;
		} );
	} );

	describe( 'Initialize', () => {
		let handleResizeHold;
		let bindEventsHold;

		beforeEach( () => {
			handleResizeHold = tribe.filterBar.filterBarSlider.handleResize;
			bindEventsHold = tribe.filterBar.filterBarSlider.bindEvents;
			tribe.filterBar.filterBarSlider.handleResize = jest.fn();
			tribe.filterBar.filterBarSlider.bindEvents = jest.fn();
		} );

		afterEach( () => {
			tribe.filterBar.filterBarSlider.handleResize = handleResizeHold;
			tribe.filterBar.filterBarSlider.bindEvents = bindEventsHold;
		} );

		test( 'Should initialize filter bar slider', () => {
			// Setup test.
			const event = {};
			const index = 0;
			const $container = $( '<div class="' + tribe.filterBar.filterBarSlider.selectors.filterBarHorizontal.className() + '"></div>' );
			const data = {};

			// Test.
			tribe.filterBar.filterBarSlider.init( event, index, $container, data );

			// Confirm final state.
			expect( tribe.filterBar.filterBarSlider.handleResize.mock.calls.length ).toBe( 1 );
			expect( tribe.filterBar.filterBarSlider.bindEvents.mock.calls.length ).toBe( 1 );
		} );

		test( 'Should return early if filter bar is not horizontal', () => {
			// Setup test.
			const event = {};
			const index = 0;
			const $container = $();
			const data = {};

			// Test.
			tribe.filterBar.filterBarSlider.init( event, index, $container, data );

			// Confirm final state.
			expect( tribe.filterBar.filterBarSlider.handleResize.mock.calls.length ).toBe( 0 );
			expect( tribe.filterBar.filterBarSlider.bindEvents.mock.calls.length ).toBe( 0 );
		} );
	} );
} );
