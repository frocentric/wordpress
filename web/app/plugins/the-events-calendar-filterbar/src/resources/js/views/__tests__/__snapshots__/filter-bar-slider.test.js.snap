// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Bar Slider Deinitialize slider Should deinitialize filter slider 1`] = `""`;

exports[`Filter Bar Slider Handle slider translate Should update filters container styles 1`] = `""`;

exports[`Filter Bar Slider Handle slider translate Should update filters container styles 2`] = `"translateX(80px)"`;

exports[`Filter Bar Slider Selectors Should match snapshot 1`] = `"{\\"filterBarHorizontal\\":\\".tribe-events--filter-bar-horizontal\\",\\"filterBar\\":\\".tribe-filter-bar\\",\\"filtersSliderContainer\\":\\"[data-js=\\\\\\"tribe-filter-bar-filters-slider-container\\\\\\"]\\",\\"filtersSliderNav\\":\\".tribe-filter-bar__filters-slider-nav\\",\\"filtersSliderNavOverflowStart\\":\\".tribe-filter-bar__filters-slider-nav--overflow-start\\",\\"filtersSliderNavOverflowEnd\\":\\".tribe-filter-bar__filters-slider-nav--overflow-end\\",\\"filtersSliderNavButtonPrev\\":\\"[data-js~=\\\\\\"tribe-filter-bar-filters-slider-nav-button-prev\\\\\\"]\\",\\"filtersSliderNavButtonNext\\":\\"[data-js~=\\\\\\"tribe-filter-bar-filters-slider-nav-button-next\\\\\\"]\\",\\"filtersContainer\\":\\"[data-js=\\\\\\"tribe-filter-bar-filters-container\\\\\\"]\\"}"`;
